MPR#7373: fix ocamlmklib when FlexDLL bootstrapped

dra27 opened the pull request:
Discussion in the [Mantis ticket](https://caml.inria.fr/mantis/view.php?id=7373).

The original problem is that when FlexDLL is bootstrapped, the runtime object files were installed to OCaml's lib directory which is fine for the compilers because, unless compiling with `-nostdlib`, these are always included. It doesn't work for ocamlmklib which doesn't include this directory by default. The effect was that ocamlmklib couldn't compile DLLs.

Rather than having ocamlmklib add the whole stdlib directory, this fix installs the FlexDLL object files to the `flexdll` subdirectory of lib. The compiler drivers automatically add `-I +flexdll` (always - so this indirectly improves the `-nostdlib` case as well) in a manner similar to how `+threads` and `+vmthreads` are added. ocamlmklib also adds an option along the lines of `-L+flexdll`, allowing DLLs to be compiled.

As a tacked on bonus, the `flexdll` bootstrapping target is also split into two so it's now possible to put pre-compiled FlexDLL object files separately (and add the appropriate -L{path} to `FLEXLINK_FLAGS`) but bootstrap just the flexlink executable as part of the build process.

dra27 committed:
Allow backslashes in FLEXLINK_FLAGS

Also in MKEXE, MKDLL and MKMAINDDLL as a result.

dra27 committed:
Install bootstrapped FlexDLL objects to flexdll/

The install-flexdll target now puts the object files for FlexDLL in a
subdirectory flexdll of the Standard Library instead of in the Standard
Library itself.

A configuration tweak means that -I +flexdll is effectively added to all
compiler invocations and also a pseudo-option -L+flexdll to ocamlmklib
calls to Config.mkdll which fixes PR#7373.

dra27 committed:
Split FlexDLL bootstrap into flexdll and flexlink

Allows the C object files to be provided separately and just the
flexlink binary itself to be bootstrapped.

dra27 commented:
Principally for the ocamlmklib fix, I'd like this to go in 4.04.1 - the fix translates onto trunk logically (some obvious conflicts in Makefiles).

@shindere - do you have a work-in-progress for merging the toplevel Makefile ... happy to comment on/provide a diff for this change.

shindere commented:
David Allsopp (2017/01/27 10:48 -0800):
> Principally for the ocamlmklib fix, I'd like this to go in 4.04.1 - the fix translates onto trunk logically (some obvious conflicts in Makefiles).
> 
> @shindere - do you have a work-in-progress for merging the toplevel
> Makefile ... happy to comment on/provide a diff for this change.

Yes. Almost done, before I had to stop working. See the
merge-root-makefiles branch of the shindere/ocaml repository on GitHub.

Please note, it's really a work in progress which is largely untested
yet. Really hope to be back soon and finish this.

dra27 commented:
I want to make a tweak to this before it's merged to split the build target off further (again, with OPAM in mind), so that installation of flexlink.exe would require `make -f Makefile.nt flexlink world.opt flexlink.opt install`. I'll do that either tomorrow or Monday and prepare a version for trunk (and 4.05?) at the same time

In the meantime, I have merged the two AppVeyor commits so that we have Windows CI properly running for 4.04.1.

dra27 committed:
Don't build flexlink.opt in opt.opt

Require flexlink.opt to be specified. Old behaviour:

make flexdll world.opt install

now becomes:

make flexdll world.opt flexlink.opt install

Without flexlink.opt, a bytecode version of flexlink.exe is installed.

dra27 committed:
Fix installation of bytecode flexlink.exe

The bytecode version of flexlink.exe is compiled before camlheader is
available, so it needs to have that prepended when it's installed.

dra27 commented:
Right, this is ready to go. Two fixes since last time:

 - (the planned one) which is that flexlink.opt must be built explicitly, so you now say `make -f Makefile.nt flexdll world.opt flexlink.opt install` to get a native version of flexlink.exe. If you omit the flexlink.opt, you still get flexlink.exe but it's...
 - bytecode installation of flexlink.exe fixed: it's compiled at a point where camlheader has not yet been compiled, so the installation needs to be prepend it.

The merging of the build systems means that this is more than just a cherry pick for 4.05 - I'll open a GPR against 4.05 once I've rebased and tested this patch on it (for testing, it will be easier for me to work against 4.05 and cherry-pick to trunk, if that's OK).

damiendoligez commented:
I am mildly disturbed by the fact that you have two `Changes` entries with the same key. Let's see what @gasche thinks of that.

dra27 commented:
Two different sections of effect! I can merge them!

I'm more concerned by the AppVeyor failure at the moment, which is to do with spaces in directory names...

gasche commented:
(I'm fine with split Changes entries, we have done this in the past I thnk.)

dra27 commented:
OK, AppVeyor was definitely doing its job there. The way I was getting ocamlmklib to pass the -L option to the command indicated by Config.mkdll (i.e. flexlink) was to hook into the machinery for -cclib. This was a big mistake because it meant that the -L option was being embedded in the .cmxa and .cma files and, with spaces in the path, they were embedded incorrectly.

I've changed it now so that the flexdll directory is passed directly to Config.mkdll, which hopefully will give an AppVeyor pass this time...

dra27 commented:
Huzzah - CI passes; I think this is safe to go.

damiendoligez commented:
Note: do not merge before the `Filename.quote` discussion in #1101 comes to a conclusion.

damiendoligez commented:
Hum. I tried to resolve the trivial conflict in `Changes` with GitHub's interface, and now it says the branch cannot be rebased because of conflicts...

@dra27 could you rebase and do the merge yourself? Thanks.

damiendoligez commented:
Cross reference: this is the 4.04 version of #1101

dra27 commented:
Note that the 4.04 version does not require the patch mentioned in https://github.com/ocaml/ocaml/pull/1101#issuecomment-291987030.

So, notwithstanding @damiendoligez's comment about always quoting, these are both fully re-tested and ready to merge.
